server/src/main/java/org/elasticsearch/index/translog/TranslogReader.java

```diff
@@ -70,6 +84,13 @@ public static TranslogReader open(
         return new TranslogReader(checkpoint, channel, path, header);
     }
 
+    /**
+     * Create a new reader with new checkoint that shares resources with current one
+     */
+    TranslogReader withNewCheckpoint(final Checkpoint newCheckpoint){
+        return new TranslogReader(newCheckpoint, channel, path, header,  closed);
```

Comment:
the are a couple of problems with this:
1) A closed translog should never be trimmed.
2) There should be a clear transfer of owner ship of the underlying channel between the old reader and the new - i.e. the old reader should be closed and the new reader should be open.

You can see how to do this kind of ownership transfer by looking at `TranslogWriter.closeIntoReader` .

Instead of splitting the trimming logic of a reader in two places (Translog.trim and here), I think we should move it all here. Call the method `closeIntoTrimmedReader` and do all the checkpoint + state juggling here.

Reply:
agreed, going to fix it